Tests/UpstreamIntegrationTests/util/AnyCodableUtils.swift

```diff
@@ -0,0 +1,848 @@
+
+                    Expected: \#(String(describing: expected))
+
+                    Actual: \#(String(describing: actual))
+
+                    Key path: \#(keyPathAsString(keyPath: keyPath))
+                """#, file: file, line: line)
+            }
+            return false
+        }
+        if expected.count != actual.count {
+            if shouldAssert {
+                XCTFail(#"""
+                    Expected and Actual counts do not match (exact equality).
+
+                    Expected count: \#(expected.count)
+                    Actual count: \#(actual.count)
+
+                    Expected: \#(expected)
+
+                    Actual: \#(actual)
+
+                    Key path: \#(keyPathAsString(keyPath: keyPath))
+                """#, file: file, line: line)
+            }
+            return false
+        }
+        var finalResult = true
+        for (key, value) in expected {
+            var keyPath = keyPath
+            keyPath.append(key)
+            finalResult = assertEqual(
+                expected: value,
+                actual: actual[key],
+                keyPath: keyPath,
+                file: file, line: line, shouldAssert: shouldAssert) && finalResult
+        }
+        return finalResult
+    }
+
+    // MARK: - AnyCodable flexible validation test assertion methods
+
+    /// Performs a flexible comparison where only the key value pairs on the expected side are required. There are two default equality modes which affect the type of validation performed.
+    ///
+    /// Given an expected JSON like the following:
+    /// {
+    /// "key1": "value1",
+    /// "key2": [{ "nest1": 1}, {"nest2": 2}]
+    /// }
+    ///
+    /// An example alternate mode path for the example JSON could be: "key2[1].nest2"
+    ///
+    /// Alternate mode paths must start from the top level of the expected JSON. Whatever key is specified by the path, from that value onward, the alternate match mode is used.
+    ///
+    /// There are 3 different ways to specify alternate mode paths for arrays:
+    /// 1. The specific index: [\<INT\>] (ex: `[0]`, `[28]`, etc.) - The element at the specified index will use the alternate mode
+    /// 2. The wildcard index: [*\<INT\>] (ex: `[*1]`, `[*12]`, etc) - The element at the specified index will use the alternate mode and apply wildcard matching logic
+    /// 3. The general wildcard: [*] (must be in exactly this format) - Every element not explicitly specified by 1 or 2 will use the alternate mode and apply wildcard matching logic. This option is mututally exclusive with default behavior.
+    /// - The default behavior is that elements from the expected JSON side are compared in order, up to the last element of the expected array
+    ///
+    /// - Parameters:
+    ///    - defaultExactEqualityMode: The default mode to use for the validation process. `true` uses exact match, and values require
```

Comment:
hmm.. it still seems confusing. What do you think about providing convenience wrappers for these asserts for more clarity, so instead of parametrized assertContains we have for e.g.:
```
assertTypeMatch(expected, actual, exactMatchPaths) -> type match, except exactMatchPaths
assertTypeMatch(expected, actual) -> all type match
assertExactMatch(expected, actual, typeMatchPaths) -> exact match, except typeMatchPaths
assertExactMatch(expected, actual) -> all exact match
assertEqual(expected, actual)
```